File: packages/oauth-resource-server-go/entitlements_test.go
package oauthrs

import (
	"context"
	"errors"
	"net/http"
	"testing"
	"time"
)

func TestEntitlementRequestVerifierRejectsReplayAcrossInstances(t *testing.T) {
	now := time.Date(2026, 8, 27, 22, 0, 0, 0, time.UTC)
	store, err := NewMemoryEntitlementReplayStore(32)
	if err != nil {
		t.Fatal(err)
	}
	store.now = func() time.Time { return now }
	config := EntitlementRequestVerifierConfig{
		SigningKey: []byte("0123456789abcdef0123456789abcdef"), Connector: "canvas", ReplayStore: store,
		Now: func() time.Time { return now },
	}
	first, err := NewEntitlementRequestVerifier(config)
	if err != nil {
		t.Fatal(err)
	}
	second, err := NewEntitlementRequestVerifier(config)
	if err != nil {
		t.Fatal(err)
	}
	body := []byte(`{"connector":"canvas","user_id":"user_1","scopes":["canvas:invoices.read"]}`)
	timestamp := now.Format(time.RFC3339)
	requestID := "0123456789abcdef0123456789abcdef"
	signature, err := SignEntitlementRequest(config.SigningKey, timestamp, requestID, body)
	if err != nil {
		t.Fatal(err)
	}
	header := make(http.Header)
	header.Set(EntitlementConnectorHeader, "canvas")
	header.Set(EntitlementTimestampHeader, timestamp)
	header.Set(EntitlementRequestIDHeader, requestID)
	header.Set(EntitlementSignatureHeader, signature)
	if err := first.Verify(context.Background(), header, body); err != nil {
		t.Fatalf("first verification failed: %v", err)
	}
	if err := second.Verify(context.Background(), header, body); !errors.Is(err, ErrEntitlementRequestReplay) {
		t.Fatalf("shared replay verification error = %v", err)
	}
}

func TestEntitlementSigningParityVectorAndTampering(t *testing.T) {
	key := []byte("0123456789abcdef0123456789abcdef")
	body := []byte(`{"allowed":true}`)
	const timestamp = "2026-08-27T22:00:00Z"
	const requestID = "request-0123456789abcdef"
	const expected = "sha256=84db9f0ae97e1b6d17b24c8b27b154a147e1abd7ccd55da3a86d764bc73f0f17"
	signature, err := SignEntitlementRequest(key, timestamp, requestID, body)
	if err != nil {
		t.Fatal(err)
	}
	if signature != expected {
		t.Fatalf("signature = %q, want parity vector %q", signature, expected)
	}

	store, _ := NewMemoryEntitlementReplayStore(8)
	store.now = func() time.Time { return time.Date(2026, 8, 27, 22, 0, 0, 0, time.UTC) }
	verifier, _ := NewEntitlementRequestVerifier(EntitlementRequestVerifierConfig{
		SigningKey: key, Connector: "canvas", ReplayStore: store, Now: store.now,
	})
	header := make(http.Header)
	header.Set(EntitlementConnectorHeader, "canvas")
	header.Set(EntitlementTimestampHeader, timestamp)
	header.Set(EntitlementRequestIDHeader, requestID+"x")
	header.Set(EntitlementSignatureHeader, signature)
	if err := verifier.Verify(context.Background(), header, body); !errors.Is(err, ErrEntitlementRequestInvalid) {
		t.Fatalf("request-ID tampering error = %v", err)
	}
}

func TestMemoryEntitlementReplayStoreFailsClosedWhenBounded(t *testing.T) {
	now := time.Now()
	store, _ := NewMemoryEntitlementReplayStore(1)
	store.now = func() time.Time { return now }
	if claimed, err := store.Claim(context.Background(), "first-request-id-0001", now.Add(time.Minute)); err != nil || !claimed {
		t.Fatalf("first claim = %v, %v", claimed, err)
	}
	if claimed, err := store.Claim(context.Background(), "second-request-id-002", now.Add(time.Minute)); err == nil || claimed {
		t.Fatalf("full bounded store claim = %v, %v", claimed, err)
	}
	now = now.Add(2 * time.Minute)
	if claimed, err := store.Claim(context.Background(), "second-request-id-002", now.Add(time.Minute)); err != nil || !claimed {
		t.Fatalf("claim after expiry = %v, %v", claimed, err)
	}
}

File: packages/oauth-resource-server-ts/test/entitlements.test.ts
import { describe, expect, it } from 'vitest';
import {
  EntitlementRequestError,
  EntitlementRequestVerifier,
  MemoryEntitlementReplayStore,
  signEntitlementRequest,
} from '../src/entitlements.js';

const key = '0123456789abcdef0123456789abcdef';
const now = Date.parse('2026-08-27T22:00:00Z');

function signedHeaders(requestId = '0123456789abcdef0123456789abcdef', body = Buffer.from('{"allowed":true}')): Record<string, string> {
  const timestamp = '2026-08-27T22:00:00Z';
  return {
    'x-rowboat-connector': 'canvas',
    'x-rowboat-timestamp': timestamp,
    'x-rowboat-request-id': requestId,
    'x-rowboat-signature': signEntitlementRequest(key, timestamp, requestId, body),
  };
}

describe('signed entitlement requests', () => {
  it('matches the Go canonical parity vector', () => {
    expect(signEntitlementRequest(
      key,
      '2026-08-27T22:00:00Z',
      'request-0123456789abcdef',
      Buffer.from('{"allowed":true}'),
    )).toBe('sha256=84db9f0ae97e1b6d17b24c8b27b154a147e1abd7ccd55da3a86d764bc73f0f17');
  });

  it('rejects replay across verifier instances sharing one bounded store', async () => {
    const store = new MemoryEntitlementReplayStore(32, () => now);
    const first = new EntitlementRequestVerifier({ signingKey: key, connector: 'canvas', replayStore: store, now: () => now });
    const second = new EntitlementRequestVerifier({ signingKey: key, connector: 'canvas', replayStore: store, now: () => now });
    const body = Buffer.from('{"allowed":true}');
    const headers = signedHeaders(undefined, body);
    await expect(first.verify(headers, body)).resolves.toBeUndefined();
    await expect(second.verify(headers, body)).rejects.toMatchObject<Partial<EntitlementRequestError>>({ code: 'replay' });
  });

  it('binds the request ID and fails closed when the bounded store is full', async () => {
    const store = new MemoryEntitlementReplayStore(1, () => now);
    const verifier = new EntitlementRequestVerifier({ signingKey: key, connector: 'canvas', replayStore: store, now: () => now });
    const body = Buffer.from('{"allowed":true}');
    const headers = signedHeaders(undefined, body);
    headers['x-rowboat-request-id'] = `${headers['x-rowboat-request-id']}x`;
    await expect(verifier.verify(headers, body)).rejects.toMatchObject<Partial<EntitlementRequestError>>({ code: 'invalid' });

    const firstHeaders = signedHeaders('first-request-id-0001', body);
    await verifier.verify(firstHeaders, body);
    const secondHeaders = signedHeaders('second-request-id-002', body);
    await expect(verifier.verify(secondHeaders, body)).rejects.toMatchObject<Partial<EntitlementRequestError>>({ code: 'store_unavailable' });
  });
});
